File: cpp/src/standard_library/standard_value_library.cpp
#include "standard_value_library.hpp"

#include "../values/ivalue.hpp"
#include "../values/array_value.hpp"
#include "../values/builtin_function_value.hpp"
#include "../virtual_machine.hpp"
#include "../scope.hpp"
#include "../utils.hpp"

namespace stack_vm
{
    std::shared_ptr<const scope> standard_value_library::scope = create_scope();

    std::shared_ptr<const scope> standard_value_library::create_scope()
    {
        auto result = std::make_shared<stack_vm::scope>();

        result->define("typeof", [](virtual_machine &vm, const array_value &args)
        {
            const auto &top = vm.pop_stack();
            vm.push_stack(top.type());
        });

            case hash("toString"):
            {
                const auto &top = vm.pop_stack();
                vm.push_stack(top.to_string());
                break;
            }
            case hash("compareTo"):
            {
                const auto &right = vm.pop_stack();
                const auto &left = vm.pop_stack();
                vm.push_stack(left.compare(right));
                break;
            }
        }

        return result;
    }
} // stack_vm

File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/VM/StandardLibrary/StandardValueLibrary.cs
using System;

namespace SimpleStackVM
{
    public static class StandardValueLibrary
    {
        #region Fields
        public const string HandleName = "value";
        #endregion

        #region Methods
        public static void AddHandler(VirtualMachine vm)
        {
            vm.AddBuiltinHandler(HandleName, Handler);
        }

        public static void Handler(string command, VirtualMachine vm)
        {
            switch (command)
            {
                case "toString":
                    {
                        var top = vm.PeekStack();
                        if (top is StringValue)
                        {
                            break;
                        }

                        top = vm.PopStack();
                        vm.PushStack(new StringValue(top.ToString()));
                        break;
                    }
                case "typeof":
                    {
                        var top = vm.PopStack();
                        vm.PushStack(new StringValue(GetTypeOf(top)));
                        break;
                    }
                case "compareTo":
                    {
                        var right = vm.PopStack();
                        var left = vm.PopStack();
                        vm.PushStack(new NumberValue(left.CompareTo(right)));
                        break;
                    }
            }
        }

        public static string GetTypeOf(IValue input)
        {
            if (input is StringValue)
            {
                return "string";
            }
            if (input is NumberValue)
            {
                return "number";
            }
            if (input is BoolValue)
            {
                return "bool";
            }
            if (input is ArrayValue)
            {
                return "array";
            }
            if (input is ObjectValue)
            {
                return "object";
            }
            if (input is NullValue)
            {
                return "null";
            }
            if (input is AnyValue)
            {
                return "any";
            }

            return "unknown";
        }
        #endregion
    }
}
